base/abstractarray.jl

```diff
@@ -2255,6 +2255,193 @@ end
 
 typed_hvcat(::Type{T}, rows::Tuple{Vararg{Int}}, as...) where T = typed_hvncat(T, rows_to_dimshape(rows), true, as...)
 
+# A fast version of hvcat for the case where we have static size information of xs
+# and the number of rows is known at compile time -- we can eliminate all the runtime
+# size checks. For cases that static size information is not beneficial, we fall back to
+# the general hvcat/typed_hvcat methods.
+#
+# Ideally, we would like to implement only `typed_hvcat_static` and let `hvcat_static` to
+# call it. However, due to the fact that Julia pack/unpacks the dynamic argument `xs` in
+# `hvcat_static(T, Val{rows}(), xs...)` with a significant overhead when `xs` is long, we
+# duplicate the code here. -- This might be improved in the future.
+#
+# One could try remove the `hvcat_static` generated function and replace
+# it with a single
+#    @inline hvcat_static(::Val{rows}, x::T, xs::Vararg{T}) where {T, rows} = typed_hvcat_static(T, Val{rows}(), x, xs...)
+# and notice the difference in performance in the following example:
+#
+# ```julia
+# julia> using BenchmarkTools
+#
+# julia> f(x, y) = [x y y y y y; y x y y y y; y y x y y y; y y y x y y; y y y y x y; y y y y y x]
+#
+# julia> @btime f(x, y) setup=(x = rand(); y = rand())
+# ```
```

Comment:
It would be nice to know how to work around the performance issue without duplicating the code.

In my local laptop, the performance is 30ns (the current duplicated generated function version) vs 900ns (without duplication)